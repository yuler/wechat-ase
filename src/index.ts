// import { createDecipheriv, createCipheriv } from 'crypto'
import { createDecipheriv } from 'crypto'

const pcks7_decode = (buf: Buffer) => {
  let padLen = buf[buf.length - 1]
  if (padLen < 1 || padLen > 32) {
    padLen = 0
  }
  return buf.slice(0, buf.length - padLen)
}

export const decode = (text: string, encodingAESKey: string) => {
  const AESKey = Buffer.from(encodingAESKey + '=', 'base64')
  const IV = AESKey.slice(0, 16)
  const decipher = createDecipheriv('aes-256-cbc', AESKey, IV)
  decipher.setAutoPadding(false)
  const decipheredBuff = pcks7_decode(
    Buffer.concat([
      decipher.update(text, 'base64'),
      decipher.final()
    ])
  )
  const messageLenngth = decipheredBuff.slice(16).slice(0, 4).readUInt32BE(0)
  const content = decipheredBuff.slice(20, messageLenngth + 20)
  return content.toString()
}