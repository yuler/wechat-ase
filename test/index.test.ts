import { encode, decode } from '../src';

describe('weixin aes', () => {
  // const token = 'shttest'
  const encodingAESKey = 'abcdefghijklmnopqrstuvwxyz0123456789ABCDEFG';
  it('encode', () => {
    const content = JSON.stringify({
      ToUserName: 'gh_0052f3068328',
      FromUserName: 'o55Go5f1BTRmw_MQMOnNDXnj6jiY',
      CreateTime: 1582701510,
      MsgType: 'text',
      Content: '1',
      MsgId: 22658835000760200,
    });
    const encrypt = encode(content, encodingAESKey);
    expect(decode(encrypt, encodingAESKey)).toEqual(content);
  });

  it('decode', () => {
    const encrypt =
      'podI1mIDNVcCql4GYcKZrM0V/KB3Y7s3k4zQCT9Apqc/IXAWxiT03mlKDSviIdp37Edlmo1QGxsIuHOiE2Pxxay2IUx/rexBTv0R6FRYI23GCbAKWnCN2p0qE+UBoHSiLSJ/Rjedk0QOfAVmC+nSCV8vqWJ7EHvL+XaX840hRNktnqe+/yp4U+F0FUHXVmNJeJq3N5ynci+1CnoKc+c6yd8hh+924cQjUHRdVnk7+DTQ/u7hyW+5SKFakBfcci47UJ1DMPmkHDE3rGcl+3pO9QCZZrcsph71EUUXxOR0lHM=';
    expect(decode(encrypt, encodingAESKey)).toMatchSnapshot();
  });
});
