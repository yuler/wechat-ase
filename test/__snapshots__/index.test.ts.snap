// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weixin aes decode 1`] = `
"{
\\"ToUserName\\": \\"gh_0052f3068328\\",
\\"FromUserName\\": \\"o55Go5f1BTRmw_MQMOnNDXnj6jiY\\",
\\"CreateTime\\": 1582701510,
\\"MsgType\\": \\"text\\",
\\"Content\\": \\"1\\",
\\"MsgId\\": 22658835000760200
}
"
`;
